ckanext/canada/view.py

```diff
@@ -228,6 +233,67 @@ def post(self):
 )
 
 
+@canada_views.route('/recover-username', methods=['GET', 'POST'])
+def recover_username():
+    if not h.is_registry() or not h.plugin_loaded('gcnotify'):
+        # we only want this route on the Registry, and the email template
+        # is monkey patched in GC Notify so we need that loaded
+        return abort(404)
+
+    context = {
+        'model': model,
+        'session': model.Session,
+        'user': g.user,
+        'auth_user_obj': g.userobj
+    }
+    try:
+        check_access('request_reset', context)
+    except NotAuthorized:
+        abort(403, _('Unauthorized to request username recovery.'))
+
+    if request.method == 'POST':
+        email = request.form.get('email')
+        if email in (None, ''):
+            h.flash_error(_('Email is required'))
+            return h.redirect_to('canada.recover_username')
+
+        log.info('Username recovery requested for email "{}"'.format(email))
+
+        context = {'model': model, 'user': g.user, 'ignore_auth': True}
+        username_list = []
+
+        user_list = get_action('user_list')(context, {'email': email})
+        if user_list:
+            for user_dict in user_list:
+                username_list.append(user_dict['name'])
+
+        if not username_list:
+            log.info('User requested username recovery for unknown email: {}'
+                     .format(email))
+        else:
+            log.info('Emailing username recovery to email: {}'
+                     .format(email))
```

Comment:
usually these things don't give away whether there are any user accounts with the email for privacy reasons. Worth thinking about with the registry accessible on the public internet at some point.

Reply:
@wardi I copied the core code from Password Reset for this, and modified a bit to just send the one email with a list of usernames.

This is just the system logging that core also had.

It will always do the flash message:
```
An email has been sent to you containing your username(s). (unless the account specified does not exist)
```
and redirect user back to home page, just like the password reset form